Kernel/FileSystem/FATFileSystem.cpp

```diff
@@ -0,0 +1,352 @@
+        dbgln("FATFS: Invalid signature");
+        return EINVAL;
+    }
+
+    m_logical_block_size = boot_record()->bytes_per_sector;
+    set_block_size(m_logical_block_size);
+
+    u32 root_directory_sectors = ((boot_record()->root_directory_entry_count * 32) + (m_logical_block_size - 1)) / m_logical_block_size;
+    m_first_data_sector = boot_record()->reserved_sector_count + (boot_record()->fat_count * boot_record()->sectors_per_fat) + root_directory_sectors;
+
+    TRY(BlockBasedFileSystem::initialize());
+
+    FATEntry root_entry {};
+    root_entry.first_cluster_low = FATInode::boot_inode_index;
+    root_entry.attributes = FATAttributes::Directory;
+    m_root_inode = TRY(FATInode::create(*this, root_entry));
+
+    return {};
+}
+
+Inode& FATFS::root_inode()
+{
+    return *m_root_inode;
+}
+
+BlockBasedFileSystem::BlockIndex FATFS::first_block_of_cluster(u32 cluster) const
+{
+    return ((cluster - 2) * boot_record()->sectors_per_cluster) + m_first_data_sector;
+}
+
+ErrorOr<RefPtr<FATInode>> FATInode::create(FATFS& fs, FATEntry entry, Vector<FATLongFileNameEntry> const& lfn_entries)
+{
+    auto filename = TRY(compute_filename(entry, lfn_entries));
+    return adopt_ref_if_nonnull(new (nothrow) FATInode(fs, entry, move(filename)));
+}
+
+FATInode::FATInode(FATFS& fs, FATEntry entry, NonnullOwnPtr<KString> filename)
+    : Inode(fs, (((u32)entry.first_cluster_high) << 16) | entry.first_cluster_low)
+    , m_entry(entry)
+    , m_filename(move(filename))
+{
+    dbgln_if(FAT_DEBUG, "FATFS: Creating inode {}", index());
+
+    m_metadata = {
+        .inode = identifier(),
+        .size = m_entry.file_size,
+        .mode = static_cast<mode_t>((has_flag(m_entry.attributes, FATAttributes::Directory) ? S_IFDIR : S_IFREG) | 0777),
+        .uid = 0,
+        .gid = 0,
+        .link_count = 0,
+        .atime = fat_date_time(m_entry.last_accessed_date, 0),
+        .ctime = fat_date_time(m_entry.creation_date, m_entry.creation_time),
+        .mtime = fat_date_time(m_entry.modification_date, m_entry.modification_time),
+        .dtime = 0,
+        .block_count = 0,
+        .block_size = 0,
+        .major_device = 0,
+        .minor_device = 0,
+    };
+}
+
+ErrorOr<Vector<BlockBasedFileSystem::BlockIndex>> FATInode::compute_block_list()
+{
+    VERIFY(m_inode_lock.is_locked());
+
+    dbgln_if(FAT_DEBUG, "FATFS: computing block list for inode {}", index());
+
+    u32 cluster = (((u32)m_entry.first_cluster_high) << 16) | m_entry.first_cluster_low;
+
+    if (cluster == boot_inode_index)
+        cluster = fs().boot_record()->root_directory_cluster;
+
+    Vector<BlockBasedFileSystem::BlockIndex> block_list;
+
+    auto fat_sector = TRY(KBuffer::try_create_with_size("FATFS: FAT read buffer"sv, fs().m_logical_block_size));
+    auto fat_sector_buffer = UserOrKernelBuffer::for_kernel_buffer(fat_sector->data());
+
+    while (cluster < 0x0FFFFFF8) {
```

Comment:
What do the magic numbers 0x0FFFFFF8 and 0x0FFFFFFF mean? Please give names to your constants. Similar to Quaker's comment on why the first two clusters are reserved.

If we have to read the FAT32 specification to understand the code, it is not descriptive enough.

Reply:
This seems like a valid critique of the code,  commenting constants isn't the serenity style. We name constants everywhere else in the codebase. This file system isn't special in that regard.

Please address this feedback.

Reply:
> I don't think that making constants for numbers used in only one place is a good idea. Instead I decided to add comments describing what do these numbers mean.

It significantly improves readability to someone that wants to poke around later on, and, as @bgianfo  has said, is in the Serenity style which dictates that there are to be no magic numbers, least of all not in the Kernel of all places.

Please fix this.

Reply:
Done.